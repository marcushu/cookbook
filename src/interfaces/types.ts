export interface Recipe {
    name: string
    owner: string
    description: string
    ingredients: string[]
    instructions: string
    breakfast: boolean
    lunch: boolean
    dinner: boolean
    glutenFree: boolean
    vegan: boolean
    vegetarian: boolean
    imageUrl: string
  }

  export interface ShoppinglistIngredient {
    ingredient: string
    recipe: string // optional to allow a user to add ingredients w/out a recipe
    /** used if the user wants the ingredients listed in a particular order */
    order?: number
  }

  export type IngredientSortPolicy = 'recipe' | 'sorting'

  export interface UserState {
    userName: string
    orderIngredientsBy: IngredientSortPolicy
    /** recipes created by this user. These exist in a different db collection
     * and need to be loaded separately. */
    recipes: Recipe[]
    favorites: Recipe[]
    shoppingList: ShoppinglistIngredient[]
  }