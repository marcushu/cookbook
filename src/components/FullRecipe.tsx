import { Box, Button, Grid, Typography } from "@mui/material";
import { FunctionComponent } from "react";
import { Recipe } from "../interfaces/types";
import ExpandLessIcon from '@mui/icons-material/ExpandLess';

interface FullRecipeProps {
  hideMe: () => void
  recipe: Recipe
}

const FullRecipe: FunctionComponent<FullRecipeProps> = ({ hideMe, recipe }) => {
  const { ingredients, instructions } = recipe;

  return (
    <Grid container py={2}>
      <Grid item xs={12} sm={4} pl={2} pb={2}>
        {ingredients.map(ingredient => 
          <Typography variant='body2' component='div' key={ingredient + recipe.name}>
            <li>
              {ingredient}
            </li>
          </Typography>)}
      </Grid>
      <Grid item xs={12} sm={8} display='flex' flexDirection='column' justifyContent='space-between'>
        <Box pl={3} pr={3}>
          {instructions.split('\n').map((sentence, index) => <Typography sx={{fontStyle: 'italic'}} key={index}>
            ~  {sentence} <br /> <br />
          </Typography>)}
        </Box>
        <Box alignSelf='end'>
          <Button onClick={hideMe} startIcon={<ExpandLessIcon fontSize='large' />} >
            less
          </Button>
        </Box>
      </Grid>
    </Grid>
  );
}

export default FullRecipe;