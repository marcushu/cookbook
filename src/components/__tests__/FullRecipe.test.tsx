import { render } from "@testing-library/react";
import { Recipe } from "../../interfaces/types";
import FullRecipe from "../FullRecipe";

describe('Fullrecipe', () => {
  const myMock = jest.fn();

  const mockRecipe = {
    ingredients: ["ing1", "ing2"],
    instructions: "this is a fake instruction"
  } as Recipe;

  it('renders', () => {
    const { asFragment } = render(<FullRecipe hideMe={myMock} recipe={mockRecipe} />)

    expect(asFragment()).toMatchSnapshot(); 
  });

  it('displays ingredients and instructions', () => {
    const { getByText } = render(<FullRecipe hideMe={myMock} recipe={mockRecipe} />);
    const ingredient = getByText(/ing1/i);
    const ingredient2 = getByText(/ing2/i);
    const instructions = getByText(/this is a fake instruction/i);

    expect(ingredient).toBeDefined();
    expect(ingredient2).toBeDefined();
    expect(instructions).toBeDefined();
  });

  it('calls prop function', () => {
    const { getByRole } = render(<FullRecipe hideMe={myMock} recipe={mockRecipe} />);

    const lessBtn =  getByRole('button');

    lessBtn.click();

    expect(myMock).toHaveBeenCalled();
  });
})


