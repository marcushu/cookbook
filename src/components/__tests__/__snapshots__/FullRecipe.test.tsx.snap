// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fullrecipe renders 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container css-1nb6n7u-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-1usvq7t-MuiGrid-root"
    >
      <div
        class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
      >
        <li>
          ing1
        </li>
      </div>
      <div
        class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
      >
        <li>
          ing2
        </li>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 css-g08sh6-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-t0051x"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-p3gdy2-MuiTypography-root"
        >
          ~  this is a fake instruction 
          <br />
           
          <br />
        </p>
      </div>
      <div
        class="MuiBox-root css-1fuqp3k"
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
              data-testid="ExpandLessIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
              />
            </svg>
          </span>
          less
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
